File: src/main/java/aurocosh/divinefavor/common/network/message/client/particles/MessageParticlesHighlight.java
package aurocosh.divinefavor.common.network.message.client.particles;

import aurocosh.divinefavor.DivineFavor;
import aurocosh.divinefavor.common.config.common.ConfigGeneral;
import aurocosh.divinefavor.common.item.talismans.spell.sense.BlockHighlighter;
import aurocosh.divinefavor.common.item.talismans.spell.sense.SenseBlockPredicate;
import aurocosh.divinefavor.common.network.base.WrappedClientMessage;
import aurocosh.divinefavor.common.util.UtilBlock;
import aurocosh.divinefavor.common.util.UtilList;
import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Vec3d;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.vecmath.Color3f;
import java.util.List;
import java.util.function.Predicate;

public abstract class MessageParticlesHighlight extends WrappedClientMessage {
    public int particles;
    public int dimensionId;
    public BlockPos position;

    public float minShift;
    public float maxShift;
    public Color3f color3f;

    public SenseBlockPredicate senseBlockPredicate;
    public String blockName;

    public MessageParticlesHighlight() {
    }

    public MessageParticlesHighlight(int particles, int dimensionId, BlockPos position, float minShift, float maxShift, Color3f color3f, SenseBlockPredicate senseBlockPredicate, String blockName) {
        this.particles = particles;
        this.dimensionId = dimensionId;
        this.position = position;
        this.minShift = minShift;
        this.maxShift = maxShift;
        this.color3f = color3f;
        this.senseBlockPredicate = senseBlockPredicate;
        this.blockName = blockName;
    }

    @Override
    @SideOnly(Side.CLIENT)
    protected void handleSafe() {
        EntityPlayer player = DivineFavor.proxy.getClientPlayer();
        World world = player.world;
        Predicate<BlockPos> predicate = getPredicate(world);
        if (predicate == null)
            return;
        if (dimensionId != world.provider.getDimension())
            return;

        List<BlockPos> positions = getHighlightPositions(player, predicate);
        List<Vec3d> vec3dList = UtilList.process(positions, Vec3d::new);
        BlockHighlighter.spawnParticles(color3f, maxShift, minShift, particles, world, vec3dList);
    }

    @SideOnly(Side.CLIENT)
    protected abstract List<BlockPos> getHighlightPositions(EntityPlayer player, Predicate<BlockPos> predicate);

    @SideOnly(Side.CLIENT)
    private Predicate<BlockPos> getPredicate(World world) {
        switch (senseBlockPredicate) {
            case BLOCK:
                Block block = Block.getBlockFromName(blockName);
                return block == null ? null : (pos -> world.getBlockState(pos).getBlock() == block);
            case WATER:
                return pos -> UtilBlock.isWater(world.getBlockState(pos).getBlock());
            case LAVA:
                return pos -> UtilBlock.isLava(world.getBlockState(pos).getBlock());
            case LIQUID:
                return pos -> UtilBlock.isLiquid(world.getBlockState(pos).getBlock());
            case ORE:
                return pos -> ConfigGeneral.ORE_BLOCKS.contains(world.getBlockState(pos).getBlock());
            case AIR:
                return world::isAirBlock;
        }
        return null;
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/network/message/client/particles/MessageParticlesHighlight.kt
package aurocosh.divinefavor.common.network.message.client.particles

import aurocosh.divinefavor.DivineFavor
import aurocosh.divinefavor.common.config.common.ConfigGeneral
import aurocosh.divinefavor.common.item.spell_talismans.sense.BlockHighlighter
import aurocosh.divinefavor.common.item.spell_talismans.sense.SenseBlockPredicate
import aurocosh.divinefavor.common.lib.extensions.getBlock
import aurocosh.divinefavor.common.lib.extensions.isLava
import aurocosh.divinefavor.common.lib.extensions.isLiquid
import aurocosh.divinefavor.common.lib.extensions.isWater
import aurocosh.divinefavor.common.lib.wrapper.ConvertingPredicate
import aurocosh.divinefavor.common.network.message.base.DivineClientMessage
import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import javax.vecmath.Color3f

abstract class MessageParticlesHighlight : DivineClientMessage {
    var particles: Int = 0
    var dimensionId: Int = 0
    var position: BlockPos = BlockPos.ORIGIN

    var minShift: Float = 0f
    var maxShift: Float = 0f
    var color3f: Color3f = Color3f(1f, 1f, 1f)

    var senseBlockPredicate: SenseBlockPredicate = SenseBlockPredicate.BLOCK
    var blockName: String = ""

    constructor()

    constructor(particles: Int, dimensionId: Int, position: BlockPos, minShift: Float, maxShift: Float, color3f: Color3f, senseBlockPredicate: SenseBlockPredicate, blockName: String) {
        this.particles = particles
        this.dimensionId = dimensionId
        this.position = position
        this.minShift = minShift
        this.maxShift = maxShift
        this.color3f = color3f
        this.senseBlockPredicate = senseBlockPredicate
        this.blockName = blockName
    }

    @SideOnly(Side.CLIENT)
    override fun handleSafe() {
        val player = DivineFavor.proxy.clientPlayer
        val world = player.world
        if (dimensionId != world.provider.dimension)
            return

        val predicate = getPredicate(world)
        val vec3dList = getHighlightPositions(player, predicate).map { Vec3d(it) }.toList()
        BlockHighlighter.spawnParticles(color3f, maxShift, minShift, particles, world, vec3dList)
    }

    @SideOnly(Side.CLIENT)
    protected abstract fun getHighlightPositions(player: EntityPlayer, predicate: (BlockPos) -> Boolean): Sequence<BlockPos>

    @SideOnly(Side.CLIENT)
    private fun getPredicate(world: World): (BlockPos) -> Boolean {
        when (senseBlockPredicate) {
            SenseBlockPredicate.BLOCK -> {
                val block = Block.getBlockFromName(blockName)
                return if (block == null) { _ -> false } else { pos -> world.getBlock(pos) === block }
            }
            SenseBlockPredicate.WATER -> return ConvertingPredicate(world::getBlockState, IBlockState::isWater)::invoke
            SenseBlockPredicate.LAVA -> return ConvertingPredicate(world::getBlockState, IBlockState::isLava)::invoke
            SenseBlockPredicate.LIQUID -> return ConvertingPredicate(world::getBlockState, IBlockState::isLiquid)::invoke
            SenseBlockPredicate.ORE -> return ConvertingPredicate(world::getBlock, ConfigGeneral.ORE_BLOCKS::contains)::invoke
            SenseBlockPredicate.AIR -> return world::isAirBlock
        }
    }
}
